pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -20,6 +21,128 @@
 from pymc_marketing.mmm.transformers import michaelis_menten
 from pymc_marketing.mmm.utils import sigmoid_saturation
 
+from pymc_marketing.mmm.components.adstock import (
+    AdstockTransformation
+)
+from pymc_marketing.mmm.components.saturation import (
+    SaturationTransformation
+)
+
+class BudgetOptimizer:
+    """
+    A class for optimizing budget allocation in a marketing mix model.
+
+    Parameters:
+    ----------
+    adstock : AdstockTransformation
+        The adstock parameter.
+    saturation : SaturationTransformation
+        The saturation parameter.
+    num_days : int
+        The number of days.
+    parameters : dict
+        A dictionary of parameters for each channel.
+    adstock_first : bool, optional
+        Whether to apply adstock transformation first or saturation transformation first. 
+        Default is True.
+
+    Methods:
+    -------
+    objective(budgets):
+        Calculate the objective function value given the budgets.
+    allocate_budget(total_budget, budget_bounds=None, custom_constraints=None):
+        Allocate the budget based on the total budget, budget bounds, and custom constraints.
+    """
+
+    def __init__(self, 
+        adstock: AdstockTransformation, 
+        saturation: SaturationTransformation, 
+        num_days: int, 
+        parameters: dict[str, dict[str, dict[str, float]]],
+        adstock_first: bool = True
+    ):
+        self.adstock = adstock
+        self.saturation = saturation
+        self.num_days = num_days
+        self.parameters = parameters
+        self.adstock_first = adstock_first
+
+    def objective(self, budgets):
+        """
+        Objective function for the allocation.
```

Comment:
Can you also give a more explicit explanation? Maybe even a simple example or motivation?